staging/src/k8s.io/apiserver/pkg/cel/openapi/compiling_test.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"testing"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/interpreter"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	apiservercel "k8s.io/apiserver/pkg/cel"
+	"k8s.io/apiserver/pkg/cel/common"
+	"k8s.io/apiserver/pkg/cel/library"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+func TestMultipleTypes(t *testing.T) {
+	env, err := buildTestEnv()
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, tc := range []struct {
+		expression         string
+		expectCompileError bool
+		expect             bool
+	}{
+		{
+			expression: "foo.foo == bar.bar",
+			expect:     true,
+		},
+		{
+			expression:         "foo.bar == 'value'",
+			expectCompileError: true,
+		},
+		{
+			expression: "foo.foo == 'value'",
+			expect:     true,
+		},
+		{
+			expression: "bar.bar == 'value'",
+			expect:     true,
+		},
+		{
+			expression: "1 + 1 == 2",
+			expect:     true,
+		},
+	} {
+		t.Run(tc.expression, func(t *testing.T) {
+			ast, issues := env.Compile(tc.expression)
+			if issues != nil {
+				if tc.expectCompileError {
+					return
+				}
+				t.Fatalf("compile error: %v", issues)
+			}
+			p, err := env.Program(ast)
+			if issues != nil {
+				t.Fatal(issues)
+			}
+			ret, _, err := p.Eval(&simpleActivation{
+				foo: map[string]any{"foo": "value"},
+				bar: map[string]any{"bar": "value"},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			if ret.Type() != types.BoolType {
+				t.Errorf("bad result type: %v", ret.Type())
+			}
+			if res := ret.Value().(bool); tc.expect != res {
+				t.Errorf("expect expression evaluates to %v, got %v", tc.expect, res)
+			}
+		})
+	}
+
+}
+
+// buildTestEnv sets up an environment that contains two variables, "foo" and
+// "bar".
+// foo is an object with a string field "foo"
+// bar is an object with a string field "bar"
```

Comment:
For testing purpose, do we want them contain different types or more fields?

Reply:
Yes we certainly do! Added more variety in the types.